Inspector: Heartbeat pull reporting

enoodle opened the pull request:
Proposed fix for #13 

cc @simon3z @pweil-

enoodle commented:
Output log:
```
$ sudo ./image-inspector --image=fedora:22 --chroot --serve=0.0.0.0:5000 --path=/tmp/image-content/ --pull-policy al
ways
2016/05/22 13:45:01 Pulling image fedora:22
2016/05/22 13:45:11 Pulling...
2016/05/22 13:45:21 Pulling...
2016/05/22 13:45:27 Extracting image fedora:22 to /tmp/image-content/
2016/05/22 13:45:34 Serving image content /tmp/image-content/ on webdav://0.0.0.0:5000/api/v1/content/
```

pweil- commented on pkg/cmd/types.go:
```diff
@@ -99,15 +113,15 @@ func (i *ImageInspectorOptions) Validate() error {
-		if !found {
-			return fmt.Errorf("%s is not one of the available scan-types which are %v", i.ScanType, ScanOptions)
+	}
+	if len(i.PullPolicy) > 0 {
```
I don't think you need the length check here.  We provide a default and if the user sets it to empty we want the error to show.

enoodle replied:
Thanks, I'll handle this and the e2e test in #15.

simon3z commented:
> ```
> 2016/05/22 13:45:11 Pulling...
> 2016/05/22 13:45:21 Pulling...
> ```

@enoodle any way to report the percentage at each log line? Thanks.

pweil- commented on cmd/image-inspector.go:
```diff
@@ -22,6 +22,7 @@ func main() {
 	flag.StringVar(&inspectorOptions.PasswordFile, "password-file", inspectorOptions.PasswordFile, "Location of a file that contains the password for authentication with the docker registry")
 	flag.StringVar(&inspectorOptions.ScanType, "scan-type", inspectorOptions.ScanType, fmt.Sprintf("The type of the scan to be done on the inspected image. Available scan types are: %v", iicmd.ScanOptions))
 	flag.StringVar(&inspectorOptions.ScanResultsDir, "scan-results-dir", inspectorOptions.ScanResultsDir, "The directory that will contain the results of the scan")
+	flag.StringVar(&inspectorOptions.PullPolicy, "pull-policy", inspectorOptions.PullPolicy, fmt.Sprintf("Pull policy, default is %s, options are: %v", iicmd.PullIfNotPresent, iicmd.PullPolicyOptions))
```
let's add an e2e test that gives an invalid policy and check for the expected error message

pweil- commented on pkg/inspector/image-inspector.go:
```diff
@@ -190,9 +195,22 @@ func (i *defaultImageInspector) pullImage(client *docker.Client) error {
 	// Try all the possible auth's from the config file
 	var authErr error
 	for _, auth := range imagePullAuths.Configs {
+		ch := make(chan bool)
```
since we're not giving any info about the specific auth being attempted does this need to live inside the for loop and create a new channel and routine for every iteration?

enoodle replied:
Right, no need to do that inside the loop. I moved this to outside and use```defer```

pweil- commented on pkg/inspector/image-inspector.go:
```diff
@@ -39,6 +40,7 @@ const (
 	OPENSCAP_URL_PATH  = API_URL_PREFIX + "/" + VERSION_TAG + "/openscap"
 	CHROOT_SERVE_PATH  = "/"
 	OSCAP_CVE_DIR      = "/tmp"
+	PULL_LOG_INTERVAL  = 10
```
let's make the name descriptive of the time unit that will be used or add a comment noting it.

enoodle replied:
changed

enoodle commented:
@simon3z The output from docker is like we see here https://gist.github.com/pweil-/9a6c4f3359d39837c657c2da40513d66 , But sometimes there are many different progress bars for all the different layers. I am not sure how to get one number to report from this.

simon3z commented:
@enoodle isn't the overall progress the average over the finite set of ongoing tasks?

enoodle commented:
@simon3z Yes, but its still very difficult because not all the values are known at the begining, and for some layers the total is never known untill its too late.
I attached a file with all this mayhem as an example, the format there is the raw json received from docker.
[docker_pull_log.zip](https://github.com/openshift/image-inspector/files/279507/docker_pull_log.zip)

simon3z commented:
> @simon3z Yes, but its still very difficult because not all the values are known at the begining, and for some layers the total is never known untill its too late.

Can we report the amount of bytes downloaded then?

enoodle commented:
@simon3z Yes, This is possible.

enoodle commented:
@simon3z I updated to print the downloaded number of bytes and updated the output example in the first comment.
@pweil- PTAL , especially the last commit. If we all like it I will squash it in.

pweil- commented on pkg/inspector/image-inspector.go:
```diff
@@ -182,6 +179,91 @@ func (i *defaultImageInspector) Inspect() error {
+			bytesDownloaded += bytes
+		case <-timeOutChannel:
+			log.Printf("Pulling... %d KB", bytesDownloaded/1024)
+			timeOutChannel = time.After(PULL_LOG_INTERVAL_SEC * time.Second)
```
Rather than creating a new channel every 10 seconds it might be more appropriate to use a ticker for this.  https://play.golang.org/p/oLwCQVXcKL

enoodle replied:
Yes, Thank you.

pweil- commented on pkg/inspector/image-inspector.go:
```diff
@@ -182,6 +179,91 @@ func (i *defaultImageInspector) Inspect() error {
+
+	// Try all the possible auth's from the config file
+	var authErr error
+	for _, auth := range imagePullAuths.Configs {
```
I'm curious about how this will display and work if there are non-auth related errors during the pull attempts.  Reading the code I would expect something like:

```
Pulling... 10 kb
Pulling... 20 kb
Pulling... 30 kb
<unseen 500 error>
Pulling... 10 kb
Pulling... 20 kb
```
We might want to display the error.

enoodle replied:
@pweil- Do you mean displaying ```err``` in line 262 if ```err != nil``` ? Or do you mean an error sent from docker? I am not sure how to catch/handle the errors from docker.

pweil- replied:
Yes, I was thinking about displaying it in this for loop.

pweil- commented:
Small items/questions.  Please add godoc to new methods.  LGTM

simon3z commented:
> 2016/05/26 15:43:00 Pulling... 22142 KB
> 2016/05/26 15:43:10 Pulling... 94817 KB
> 2016/05/26 15:43:20 Pulling... 196780 KB
> 2016/05/26 15:43:30 Pulling... 258544 KB
> 2016/05/26 15:43:40 Pulling... 347079 KB
> 2016/05/26 15:43:50 Pulling... 407236 KB

Minor... I'd go with something more explicit:

```
2016/05/26 15:43:00 Pulling image (22142Kb downloaded)
```

or

```
2016/05/26 15:43:00 Downloading image (22142Kb downloaded)
```

Or anything nicer.

enoodle commented on pkg/inspector/image-inspector.go:
```diff
@@ -177,25 +179,94 @@ func (i *defaultImageInspector) Inspect() error {
 		if authErr = client.PullImage(imagePullOption, auth); authErr == nil {
 			return nil
 		}
+		log.Printf("Authentication with %s failed: %v", name, authErr)
```
@pweil- I added this to print authentication errors, is that what you were aiming for?

pweil- replied:
Yes, thanks.   That will at least log something for us if the PullImage has an error that is not auth related.

cben commented on pkg/inspector/image-inspector.go:
```diff
@@ -182,6 +179,98 @@ func (i *defaultImageInspector) Inspect() error {
+		ProgressDetail progressDetailType
+	}
+	defer func() { close(bytesChan) }()           // Closing the channel to end the other routine
+	layersByetsDownloaded := make(map[string]int) // bytes downloaded per layer
```
s/byets/bytes/g

enoodle commented:
To answer some points that were raised last week by @simon3z and @cben:
- When a layer finishes downloading we have a "Downloading" status with the full downloaded value of that layer so the current aggregation is correct.
- If the routine that reads from the pipe will crush for some reason then the pipe might get blocked for reading (when the buffer will be full) and it might block the docker client (depending on how it is written). On the other hand this function is pretty simple - The only function call is to ```dec.Decode``` and we check its error value.
I think maybe add a small loop like:
```
for err := reader.Read(); err != nil {
}
```
to keep the pipe from blocking. What do you think?

simon3z commented:
@enoodle can you fix this up? It seems interesting to have.

simon3z commented:
@enoodle what's the status of this?

@pweil- do you want to review?

simon3z commented:
@enoodle can you remove the pull-policy patch that doesn't belong here. Thanks.

enoodle commented:
@simon3z removed

enoodle commented:
@simon3z This is ready to be merged.

enoodle commented:
Fixes #13
